### PR TITLE
Enforce `discovery.zen.minimum_master_nodes` is set when bound to a public ip

### PR DESCRIPTION
**Pull request by bleskes**
discovery.zen.minimum_master_nodes is the single most important setting to set on a production cluster. We have no way of supplying a good default so it must be set by the user. Binding a node to a public IP (as opposed to the default local host) is a good enough indication that a node will be part of a production cluster cluster and thus it's a good tradeoff to enforce the settings. Note that nothing prevent users from setting it to 1 in a single node cluster.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 1]
```diff
@@ -135,9 +137,10 @@ static boolean enforceLimits(final Settings settings) {
         /**
          * test if the node fails the check
          *
+         * @param settings current node settings, which may influence the check
          * @return true if the node failed the check
          */
-        boolean check();
+        boolean check(Settings settings);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: I don't like that settings need to be passed into every check method. I'd prefer the specific setting needed be passed into `MinMasterNodesCheck` when it's constructed (look at the `MlockallCheck`).

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 2]
```diff
@@ -224,12 +227,25 @@ boolean isMemoryLocked() {
 
     }
 
+    static class MinMasterNodesCheck implements Check {
+
+        @Override
+        public boolean check(Settings settings) {
+            return settings.get("discovery.zen.minimum_master_nodes") == null;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: Use `ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.exists(settings)`?

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 3]
```diff
@@ -184,10 +187,10 @@ public final boolean check() {
         @Override
         public final String errorMessage() {
             return String.format(
-                    Locale.ROOT,
-                    "max file descriptors [%d] for elasticsearch process likely too low, increase to at least [%d]",
-                    getMaxFileDescriptorCount(),
-                    limit
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: Unrelated formatting changes: :cry:

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 4]
```diff
@@ -224,12 +227,25 @@ boolean isMemoryLocked() {
...
+    static class MinMasterNodesCheck implements Check {
+
+        @Override
+        public boolean check(Settings settings) {
+            return settings.get("discovery.zen.minimum_master_nodes") == null;
+        }
+
+        @Override
+        public String errorMessage() {
+            return "please set [discovery.zen.minimum_master_nodes] to a majority of the number of master eligible nodes in your cluster.";
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: Can you replace the string `discovery.zen.minimum_master_nodes` with `ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTINGS.getKey()`?

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java [hunk 5]
```diff
@@ -197,4 +199,15 @@ public void testEnforceLimits() {
...
     }
 
+    public void testMinMasterNodes() {
+        BootstrapCheck.Check check = new BootstrapCheck.MinMasterNodesCheck();
+        assertTrue(check.check(Settings.EMPTY));
+        assertFalse(check.check(
+            Settings.builder().put(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey(), 1).build()
+        ));
+        expectThrows(RuntimeException.class, () ->
+            BootstrapCheck.check(
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: There's a `Bootstrap#check(boolean, Settings)` override that is visible for testing exactly so that the test can be written without having to rely on passing in a setting the triggers enforcement.

**Comment by jasontedor** [comment]
There's a comment at the top of `BootstrapCheck.java` about `discovery.zen.minimum_master_nodes`; can you remove it now?

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java (hunk 5 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: I wanted to check that the default set of checks enforces this. shall I add BoostrapCheck.check(boolean, settings)?

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: Okay, that's probably not a bad test to have; I think it'd be better to make `BootstrapCheck#checks(Settings)` package visible (and final) and check that `checks(Settings.EMPTY)` contains an instance of `MinMasterNodesCheck` (it's a more direct test of what you're trying to test).

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: I personally wanted to check "outside behavior" without really know what the thing does. I give a setting with no min master nodes and strictness is enforced -> boom.  I don't care how. Here too -I don't mind much. Can change if you prefer.

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 6]
```diff
@@ -93,12 +93,12 @@ static void check(final boolean enforceLimits, final List<Check> checks) {
     // visible for testing
     static Set<Setting> enforceSettings() {
         return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
-            TransportSettings.BIND_HOST,
-            TransportSettings.HOST,
-            TransportSettings.PUBLISH_HOST,
-            NetworkService.GLOBAL_NETWORK_HOST_SETTING,
-            NetworkService.GLOBAL_NETWORK_BINDHOST_SETTING,
-            NetworkService.GLOBAL_NETWORK_PUBLISHHOST_SETTING
+                TransportSettings.BIND_HOST,
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: More unrelated formatting changes? Disable auto-format on commit?

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 7]
```diff
@@ -93,12 +93,12 @@ static void check(final boolean enforceLimits, final List<Check> checks) {
     // visible for testing
     static Set<Setting> enforceSettings() {
         return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
-            TransportSettings.BIND_HOST,
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: More unrelated formatting changes? Disable auto-format on commit?

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 8]
```diff
@@ -236,11 +255,11 @@ public boolean check() {
         @Override
         public String errorMessage() {
             return String.format(
-                Locale.ROOT,
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: More unrelated formatting changes? Disable auto-format on commit?

File: core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java [hunk 9]
```diff
@@ -260,10 +279,10 @@ public boolean check() {
...
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java]: More unrelated formatting changes? Disable auto-format on commit?

File: docs/reference/migration/migrate_5_0/settings.asciidoc [hunk 10]
```diff
@@ -201,3 +201,11 @@ setting settings via `--name.of.setting value.of.setting`. This feature
 has been removed. Instead, use
 `-Ees.name.of.setting=value.of.setting`. Note that in all cases the
 name of the setting must be prefixed with `es.`.
+
+==== Discovery Settings
+
+The `discovery.zen.minimum_master_node` must bet set for nodes that are bound
+to a public network interface. We see those nodes as in "production" mode and
```
Comment (jasontedor) [inline docs/reference/migration/migrate_5_0/settings.asciidoc]: `to a public` -> `to a non-loopback`

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java (hunk 5 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: Yeah, let's the keep the tests just focused on whether or not `MinMasterNodeCheck` does the right thing based on whether or not `discovery.zen.minimum_master_nodes` is set and we can think about broader tests for the default checks from `BootstrapCheck` itself in a separate pull request.

**Comment by jasontedor** [comment]
@bleskes Thank you for picking this one up. I left a few more comments about the unrelated formatting changes and the test; feel free to process and push at your discretion.

It would be nice to someday make the check much stronger than whether or not the setting is set at all (i.e., enforce that it's a quorum). I think that we should think about this over the next few months as we work on related problems in this area.

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java (hunk 5 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapCheckTests.java]: > MinMasterNodeCheck does the right thing based on whether or not discovery.zen.minimum_master_nodes is set 

The problem is that now that we moved the settings check out of MinMasterNodeCheck and into the code that calls it's constructor, I need to make sure it's created correctly and does the right thing. I changed the test to create the default checks and pass them to the check methods.
